src/java.base/share/classes/java/util/zip/ZipUtils.java

```diff
@@ -170,23 +171,31 @@ static LocalDateTime javaEpochToLocalDateTime(long time) {
      * The bytes are assumed to be in Intel (little-endian) byte order.
      */
     public static final int get16(byte[] b, int off) {
-        return (b[off] & 0xff) | ((b[off + 1] & 0xff) << 8);
+        Preconditions.checkIndex(off, b.length, Preconditions.AIOOBE_FORMATTER);
+        Preconditions.checkIndex(off + 1, b.length, Preconditions.AIOOBE_FORMATTER);
+        return Short.toUnsignedInt(
+                UNSAFE.getShortUnaligned(b, off + Unsafe.ARRAY_BYTE_BASE_OFFSET, false));
     }
 
     /**
      * Fetches unsigned 32-bit value from byte array at specified offset.
      * The bytes are assumed to be in Intel (little-endian) byte order.
      */
     public static final long get32(byte[] b, int off) {
-        return (get16(b, off) | ((long)get16(b, off+2) << 16)) & 0xffffffffL;
+        Preconditions.checkIndex(off, b.length, Preconditions.AIOOBE_FORMATTER);
+        Preconditions.checkIndex(off + 3, b.length, Preconditions.AIOOBE_FORMATTER);
+        return Integer.toUnsignedLong(
+                UNSAFE.getIntUnaligned(b, off + Unsafe.ARRAY_BYTE_BASE_OFFSET, false));
     }
 
     /**
      * Fetches signed 64-bit value from byte array at specified offset.
      * The bytes are assumed to be in Intel (little-endian) byte order.
      */
     public static final long get64(byte[] b, int off) {
```

Comment:
This method returns a signed 64 bit value, which I think is not what some of its call sites expect. It should in any case be renamed to `get64S` to align with `get32S`. A new method `get64` should be introduced and any call site expecting unsigned numbers (most?) should use that instead.

If you don't want to deal with this in this PR, I could file an issue and suggest a PR for this. Let me know.

Reply:
As it's a pre-existing issue I'd prefer to keep this one focused on the switch-over. How would you model unsigned long values here, though? Sure we could read into a `BigInteger` but to really support such overflows we might have to rework a lot of things. 

We already cap some values even lower in practice:
```
                            end.centot = (int)centot64; // assume total < 2g
```

Reply:
> How would you model unsigned long values here, though?

I don't think we should. `9223372036854775807 ` should be enough for everyone :-)

It may be worth renaming the method to `get64S` and add a `get64` variant which either clamps at `LONG.MAX_VALUE` or throws `IllegalArgumentException` for larger values. Call sites doing custom validation `checkZip64ExtraFieldValues` could then call `get64S` and check for a negative long.

But that's food for another PR.

Reply:
Renaming to `get64S` is reasonable to be internally consistent. Updated. Improving validation of data in such 64-bit fields I'll leave for the future. I think a reasonable stance is to throw in the `check` methods if any such field is negative, at least for some of these fields.

Reply:
> FWIW we already cap some values even lower in practice:
> 
> ```
>                             end.centot = (int)centot64; // assume total < 2g
> ```

I submitted #21384 which adds validation of end.centot and also eliminates this narrowing conversion.